File: windows-app/Plantoir.Core/Models/GradedFolderRule.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Plantoir.Core.Models;

/// <summary>
/// Which folders hold work that COUNTS FOR MARKS.
///
/// <para>An expectation is "assessed" — the ring on a cell in the Curriculum
/// Coverage map, and Ontario's requirement that every overall expectation be
/// evaluated at least once — when a page addressing it lives in one of these.
/// It used to be hardcoded: any folder whose name CONTAINED "task". A teacher
/// who called theirs "Tests", or renamed "Tasks", silently lost every assessed
/// mark on the map with nothing said.</para>
///
/// <para><b>Absent is not empty, and conflating the two breaks courses both
/// ways.</b> An ABSENT <c>graded_folders</c> key means the teacher has never
/// been asked, so the historical substring rule still applies. An EMPTY list
/// means they were asked and cleared it, which is a real answer and is
/// honoured. The exact-name rule is NARROWER than the substring one —
/// <c>support/skeletons</c> ships a family whose folder is "Thinking Tasks",
/// which the old rule counted and a pool of ["Tasks"] does not — so seeding
/// every existing course with ["Tasks"] would have quietly taken the assessed
/// marks off that course's map.</para>
///
/// <para>Pinned by <c>contracts/shared-rules.json</c> -> <c>gradedFolders</c>,
/// whose cases the macOS suite, <c>scripts/test_graded_folders.py</c> and
/// <c>GradedFolderContractTests</c> all run against their own
/// implementations.</para>
/// </summary>
public static class GradedFolderRule
{
    /// <summary>
    /// The substring a folder's name must contain for a course that has never
    /// been asked which folders count.
    /// </summary>
    public const string HistoricalSubstring = "task";

    /// <summary>
    /// Whether a page counts for marks.
    ///
    /// <param name="pooledNames">
    /// The course's <c>graded_folders</c>, or <b>null</b> when the key is
    /// absent — which is not the same as an empty list.
    /// </param>
    /// <param name="relativePath">
    /// The page's path RELATIVE to the content root. Either separator.
    /// </param>
    /// </summary>
    public static bool CountsForMarks(IReadOnlyList<string>? pooledNames, string relativePath)
    {
        if (string.IsNullOrWhiteSpace(relativePath)) return false;

        var segments = relativePath.Split(new[] { '/', '\\' }, StringSplitOptions.RemoveEmptyEntries);
        // The last segment is the FILE NAME, and a page is not assessed work
        // on account of what it is called. Folder segments only, at any depth,
        // so a teacher who files by unit inside Tasks does not lose credit.
        if (segments.Length < 2) return false;
        var folders = segments[..^1];

        if (pooledNames is null)
            return folders.Any(f => f.Contains(HistoricalSubstring, StringComparison.OrdinalIgnoreCase));

        var pool = new HashSet<string>(
            pooledNames.Where(n => !string.IsNullOrEmpty(n)),
            StringComparer.OrdinalIgnoreCase);
        return folders.Any(pool.Contains);
    }

    /// <summary>
    /// The pool a course that has never been asked is ALREADY working to,
    /// read off the folders it actually has.
    ///
    /// <para>Used to materialise a <c>null</c> pool the first time a teacher
    /// edits it, so their first tick does not silently narrow the course from
    /// the substring rule to whatever one box they touched. Mirrors the
    /// fallback branch of <c>setup_course.py:graded_folders_for</c>.</para>
    /// </summary>
    public static List<string> InferredPool(IEnumerable<string>? folderNames)
    {
        var found = new List<string>();
        foreach (string name in folderNames ?? Enumerable.Empty<string>())
        {
            if (string.IsNullOrEmpty(name)) continue;
            if (!name.Contains(HistoricalSubstring, StringComparison.OrdinalIgnoreCase)) continue;
            if (found.Contains(name, StringComparer.Ordinal)) continue;
            found.Add(name);
        }
        return found;
    }

    /// <summary>
    /// A declared pool narrowed to the folders the course actually ends up
    /// with, keeping the course's own capitalisation.
    ///
    /// <para>The wizard's half of <c>setup_course.py:graded_folders_for</c>: a
    /// name the teacher removed while setting the course up is dropped rather
    /// than written into a pool that matches nothing on disk.</para>
    /// </summary>
    public static List<string> Reconciled(IEnumerable<string>? declared, IEnumerable<string>? actualFolders)
    {
        var actual = (actualFolders ?? Enumerable.Empty<string>())
            .Where(n => !string.IsNullOrEmpty(n)).ToList();
        var byLowerName = new Dictionary<string, string>(StringComparer.Ordinal);
        // LAST wins, not first: setup_course.py builds this with a dict
        // comprehension (`{name.lower(): name for name in actual_list}`),
        // where a later key overwrites an earlier one. TryAdd would keep the
        // FIRST and disagree with the build for a course holding both
        // "Tasks" and "tasks".
        foreach (string name in actual)
            byLowerName[name.ToLowerInvariant()] = name;

        var reconciled = new List<string>();
        foreach (string name in declared ?? Enumerable.Empty<string>())
        {
            if (string.IsNullOrEmpty(name)) continue;
            string? target = actual.Contains(name, StringComparer.Ordinal)
                ? name
                : byLowerName.TryGetValue(name.ToLowerInvariant(), out var matched) ? matched : null;
            if (target is not null && !reconciled.Contains(target, StringComparer.Ordinal))
                reconciled.Add(target);
        }
        return reconciled;
    }

    // ---- What a teacher reads on the Marks control ------------------------

    /// <summary>
    /// The title of the tick list, and the caption below it. Pinned by
    /// <c>contracts/shared-rules.json</c> -> <c>gradedFolders.wording</c>,
    /// proposed from Windows 2026-09-08.
    ///
    /// <para>The same two strings serve Course Settings AND the New Course
    /// Wizard, on both platforms. Before this they were four different
    /// strings pinned by nothing: this app said "Folders that count for
    /// marks" and named the map "the Curriculum Coverage map", which is the
    /// built PAGE's title used where a common noun belongs.</para>
    ///
    /// <para><b>The caption belongs BELOW its list.</b> It says "a page in one
    /// of these", and this app used to draw it ABOVE, where "these" followed
    /// the section header "Marks" and referred to nothing. Moving it is what
    /// lets one string serve all four surfaces.</para>
    /// </summary>
    public const string ListTitle = "Folders whose work counts for marks";

    /// <summary>
    /// Says "tick", not "add" or "remove": this is a tick list with no Add
    /// button, and unticking the last graded folder while the coverage map is
    /// on is refused outright (<see cref="SpecialNames.LastGradedFolderBlocked"/>).
    /// </summary>
    public const string Caption =
        "Tick the folders holding work that counts for marks. The curriculum coverage map shows an expectation as evaluated when a page in one of these addresses it. Most courses keep “Tasks”; tick “Tests” or anything else you mark.";
}

File: mac-app/QuartzTeachers/Models/GradedFolderRule.swift
import Foundation

/// The rule a course is ALREADY working to when nobody has ever been asked
/// which folders count for marks: any folder whose name mentions tasks.
///
/// `graded_folders` absent in `course_config.json` does not mean "none" — it
/// means the question was never put, so the historical substring rule still
/// applies and the build counts what it always counted. Three places in this
/// app had to answer "what is this course counting right now?" and each had
/// written the substring out for itself; a fourth copy arrived with the
/// wizard's skeleton toggle, which is why the rule is now in one place.
///
/// Windows' half is `Plantoir.Core/Models/GradedFolderRule.cs`; the name here
/// matches its `InferredPool` deliberately, because
/// `contracts/shared-rules.json` compares the two apps' answers rather than
/// their code.
enum GradedFolderRule {

    // MARK: - Stored properties

    /// What a folder's name has to contain, case ignored, for the build to
    /// have been counting it before anybody was asked. Mirrors
    /// `setup_course.py`'s `graded_folders_for` fallback and
    /// `build_site.py`'s own test.
    nonisolated static let historicalSubstring: String = "task"

    // MARK: - Functions

    /// A pool somebody CHOSE — ticked in the wizard, or declared by a
    /// payload's manifest — reconciled against the folders the course
    /// actually ends up with, in the order it was chosen.
    ///
    /// A name they ticked and then took out of the course would otherwise be
    /// written into `graded_folders` matching nothing on disk: the file would
    /// say something untrue, and the two apps would write DIFFERENT files for
    /// the same clicks. (There is no second net: since GitHub issue #192
    /// `setup_course.py` writes a saved pool back as it was, so what the
    /// wizard writes is what the course keeps — contracts/shared-rules.json
    /// → `gradedFolders.rerunningSetup`.)
    ///
    /// **The command line's rule, to the letter** —
    /// `setup_course.graded_folders_for`, and Windows'
    /// `GradedFolderRule.Reconciled`, which already followed it: a blank
    /// goes; a name that is a folder EXACTLY is kept as written; otherwise a
    /// name that is a folder with case ignored takes the folder's own
    /// spelling — the LAST folder of that spelling when two differ only by
    /// case, as a Python dictionary built in order gives; any other name is
    /// dropped; and a name already kept is not kept again. Pinned by
    /// `gradedFolders.reconcilingAChosenPool`, which all three run.
    ///
    /// Until [issue #152](https://github.com/russellgordon/plantoir/issues/152)
    /// (from #85's second item) this matched EXACTLY and kept repeats, so a
    /// `tasks` chosen against a folder `Tasks` wrote an empty pool where the
    /// command line writes `["Tasks"]`. No shipped payload or skeleton shows
    /// the difference — every pool matches its own folders exactly — so this
    /// is one rule in one place, shared with `ExampleContentCatalog.marksPool`,
    /// rather than a fix a teacher will notice.
    nonisolated static func reconciled(_ declared: [String], toFolders folders: [String]) -> [String] {
        // The later folder wins the spelling, as it does in Python's
        // dictionary built the same way.
        var spellingIgnoringCase: [String: String] = [:]
        for folder in folders {
            if folder.isEmpty {
                continue
            }
            spellingIgnoringCase[folder.lowercased()] = folder
        }

        var kept: [String] = []
        for name in declared {
            if name.isEmpty {
                continue
            }
            var folderName: String? = nil
            if folders.contains(name) {
                folderName = name
            } else if let respelled = spellingIgnoringCase[name.lowercased()] {
                folderName = respelled
            }
            guard let matchedFolder = folderName else {
                continue
            }
            if !kept.contains(matchedFolder) {
                kept.append(matchedFolder)
            }
        }
        return kept
    }

    /// The pool a course that has never been asked is already working to,
    /// read off the folders it actually has — in the order they were given,
    /// each name once.
    ///
    /// De-duplicated by exact name, matching Windows' `InferredPool`: the
    /// same folder can reach this from two lists at once (a course's shared
    /// folders and its per-section folders), and offering a teacher the same
    /// name twice would tick one box and leave the other looking unticked.
    nonisolated static func inferredPool(from folderNames: [String]) -> [String] {
        var counted: [String] = []
        for name in folderNames {
            if name.isEmpty {
                continue
            }
            if !name.lowercased().contains(historicalSubstring) {
                continue
            }
            if counted.contains(name) {
                continue
            }
            counted.append(name)
        }
        return counted
    }
}
